OpenRA.Mods.Common/Effects/LineImpactProjectile.cs

```diff
@@ -0,0 +1,271 @@
+				var maxOffset = inaccuracy * (target - headPos).Length / args.Weapon.Range.Length;
+				target += WVec.FromPDF(world.SharedRandom, 2) * maxOffset / 1024;
+			}
+
+			facing = OpenRA.Traits.Util.GetFacing(target - headPos, 0);
+
+			// Update the target position with the range we shoot beyond the target by
+			// I.e. we can deliberately overshoot, so aim for that position
+			var dir = new WVec(0, -1024, 0).Rotate(WRot.FromFacing(facing));
+			var overshoot = dir * info.BeyondTargetRange.Length / 1024;
+			target = target + overshoot;
+
+			length = Math.Max((target - headPos).Length / speed.Length, 1);
+
+			if (args.SourceActor.IsInWorld)
+				offSet = args.Source - args.SourceActor.CenterPosition;
+		}
+
+		public void Tick(World world)
+		{
+			headTicks++;
+			if (headTicks >= length)
+			{
+				headPos = target;
+				headTravelling = false;
+			}
+			else
+			{
+				if (headTravelling)
+					headPos = WPos.LerpQuadratic(args.Source, target, angle, headTicks, length);
+			}
+
+			// While the head is travelling, the tail must start to follow BeamLengthDuration ticks later.
+			// Alternatively, also stop emitting the beam if source actor dies or is ordered to stop.
+			if ((headTicks >= info.BeamFireDuration && !tailTravelling) || args.SourceActor.IsDead || args.SourceActor.IsIdle)
+				tailTravelling = true;
+
+			if (tailTicks <= 0)
+			{
+				if (args.SourceActor.IsInWorld && !args.SourceActor.IsDead)
+				{
+					args.Source = args.SourceActor.CenterPosition + offSet;
+					tailPos = args.Source;
+				}
+			}
+
+			if (tailTravelling)
+			{
+				tailTicks++;
+				if (tailTicks >= length)
+				{
+					tailPos = target;
+					tailTravelling = false;
+				}
+				else
+				{
+					tailPos = WPos.LerpQuadratic(args.Source, target, angle, tailTicks, length);
+				}
+			}
+
+			// Raycast to find all affected actors
+			var hitActors = world.FindActorsOnLine(tailPos, headPos, info.BeamWidth, info.TargetExtraSearchRadius);
+
+			// If this projectile is blockable, first check if any of the hit actors will block this.
+			if (info.Blockable)
+			{
+				var wasBlocked = false;
+				foreach (var victim in hitActors)
+				{
+					if (victim.TraitsImplementing<BlocksProjectiles>().Any(Exts.IsTraitEnabled))
+					{
+						// Find where on the line this actor blocked the beam, and then limit the beam to that point.
+						var projectedPoint = WorldUtils.MinimumPointLineProjection(tailPos, headPos, victim.CenterPosition);
+						if ((projectedPoint - tailPos).Length < (headPos - tailPos).Length)
+						{
+							headPos = projectedPoint;
+							wasBlocked = true;
+						}
+					}
+				}
+
+				if (wasBlocked)
+				{
+					target = headPos;
+					length = Math.Min(headTicks, length);
+					hitActors = world.FindActorsOnLine(headPos, tailPos, info.BeamWidth, info.TargetExtraSearchRadius);
+				}
+			}
+
+			if (headTicks % info.TicksBetweenImpacts == 0)
+			{
+				var falloff = 100;
```

Comment:
Variables should be restricted to the minimum scope in which they are used.  Remove this line, and instead declare falloff the first time that you assign to it in the inner loop.